src/eventDefinitions/LinkEventDefinition.js

```diff
@@ -0,0 +1,143 @@
+
+    debug(`<${executionId} (${id})> expect ${description}`);
+
+    broker.publish('event', 'activity.wait', {
+      ...messageContent,
+      executionId: parentExecutionId,
+      parent: shiftParent(parent),
+      link: {...reference},
+    });
+
+    function onCatchLink(routingKey, message) {
+      if (getPropertyValue(message, 'content.message.linkName') !== reference.linkName) return;
+      if (message.content.state === 'discard') return discard();
+      return complete('caught', message.content.message);
+    }
+
+    function onApiMessage(routingKey, message) {
+      const messageType = message.properties.type;
+
+      switch (messageType) {
+        case 'link': {
+          return complete('got link with', message.content.message);
+        }
+        case 'discard': {
+          return discard();
+        }
+        case 'stop': {
+          stop();
+          break;
+        }
+      }
+    }
+
+    function complete(verb, output) {
+      completed = true;
+
+      stop();
+
+      debug(`<${executionId} (${id})> ${verb} ${description}`);
+
+      broker.publish('event', 'activity.catch', {
+        ...messageContent,
+        link: {...reference},
+        message: {...output},
+        executionId: parentExecutionId || executionId,
+        parent: shiftParent(executeMessage.content.parent),
+      }, {type: 'catch'});
+
+      return broker.publish('execution', 'execute.completed', {...messageContent, output, state: 'catch'});
+    }
+
+    function discard() {
+      completed = true;
+      stop();
+      return broker.publish('execution', 'execute.discard', {...messageContent});
+    }
+
+    function stop() {
+      broker.cancel(`_api-link-${executionId}`);
+      broker.cancel(`_api-parent-${parentExecutionId}`);
+      broker.cancel(`_api-${executionId}`);
+      broker.purgeQueue(linkQueueName);
+    }
+  }
+
+  function executeThrow(executeMessage) {
+    const messageContent = cloneContent(executeMessage.content);
+    const {executionId, parent} = messageContent;
+    const parentExecutionId = parent && parent.executionId;
+
+    const description = messageDescription();
+
+    debug(`<${executionId} (${id})> throw ${description}`);
+
+    broker.publish('event', 'activity.link', {
+      ...cloneContent(messageContent),
+      executionId: parentExecutionId,
+      parent: shiftParent(parent),
+      message: {...reference},
+      state: 'throw',
+    }, {type: 'link', delegate: true});
+
+    return broker.publish('execution', 'execute.completed', messageContent);
+  }
+
+  function messageDescription() {
+    return `link ${reference.linkName}`;
+  }
+
+  function setupCatch() {
+    broker.assertQueue(linkQueueName, {autoDelete: false, durable: true});
+    broker.bindQueue(linkQueueName, 'api', '*.link.#', {durable: true});
+  }
+
+  function setupThrow() {
+    broker.subscribeTmp('event', 'activity.discard', onDiscard, {noAck: true, consumerTag: '_link-parent-discard'});
```

Comment:
I see from your changes here that I missed some things with respect to the activity lifecycle.  Is that lifecycle documented somewhere?  It would be helpful for making changes in the future.

Reply:
Thank you for the review. Hope it still works as expected after my refactoring.

It's not really about the lifecycle, which is not documented, sorry about that. The issue is that the execution stall if a link throw event is discarded.

Background:
- An activity without inbound sequence flows, unless boundary, is considered a start activity
- The process will initialise all start activities on execute
- The behaviour execute function is not called if an activity is discarded

Hence, a catch link event will be started waiting for a link message. If the link throw event is discarded the link message is never fired, since the behaviour execute function is not called. Then the catch link event waits indefinitely.

I solved it by letting the link event behaviour listen for broker discard message. If so, fire a delegate link message with a discard flag. Presto!

Perhaps not the most elegant solution but it works for now.

Reply:
> An activity without inbound sequence flows, unless boundary, is considered a start activity

Okay, I hadn't realized that's what was happening here, that actually explains a lot - thanks!

Reply:
Added [activity lifecycle documentation](/paed01/bpmn-elements/blob/master/docs/Activity.md).